libresin/libresin/core/resources/shader_resource.hpp

```diff
@@ -0,0 +1,96 @@
+#ifndef SHADER_RESOURCE_HPP
+#define SHADER_RESOURCE_HPP
+
+#include <array>
+#include <cstdint>
+#include <filesystem>
+#include <libresin/core/resources/resource_manager.hpp>
+#include <libresin/utils/exceptions.hpp>
+#include <string_view>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace resin {
+
+namespace shader_macros {
+
+static constexpr std::string_view kSupportedShaderVersion = "#version 150\n";
+
+static constexpr std::string_view kIncludeMacro = "#include";
+static constexpr std::string_view kExtDefiMacro = "#external_definition";
+static constexpr std::string_view kVersionMacro = "#version";
```

Comment:
What plans do you have regarding `#version` as a macro?

Reply:
Currently it's defined here in order to detect version macros during preprocessing and remove them as we are forcing `#version 150` at the top of the shader. From [khronos](https://www.khronos.org/opengl/wiki/Core_Language_(GLSL))
> The #version directive must appear before anything else in a shader, save for whitespace and comments. If a #version directive does not appear at the top, then it assumes 1.10, which is almost certainly not what you want. 

This way we avoid redundant `#version` macros in the resultant shader.

Future plans are rather the lowest priority. We could check the dependency version and add e.g. `DependencyIncompatibleVersionException` and in the case of success put version macro specified in the `.frag`/`.vert` at the top.